components/graph/git-ref.js

```diff
@@ -76,58 +76,54 @@ RefViewModel.prototype.dragEnd = function() {
+      promise = this.server.postPromise('/reset', { path: this.graph.repoPath(), to: target, mode: 'hard' });
     } else if (this.isTag) {
-      this.server.post('/tags', { path: this.graph.repoPath(), name: this.refName, sha1: target, force: true }, callbackWithRefSet);
+      promise = this.server.postPromise('/tags', { path: this.graph.repoPath(), name: this.refName, sha1: target, force: true });
     } else {
-      this.server.post('/branches', { path: this.graph.repoPath(), name: this.refName, sha1: target, force: true }, callbackWithRefSet);
+      promise = this.server.postPromise('/branches', { path: this.graph.repoPath(), name: this.refName, sha1: target, force: true });
     }
   } else {
     var pushReq = { path: this.graph.repoPath(), remote: this.remote, refSpec: target, remoteBranch: this.refName };
-    this.server.post('/push', pushReq, function(err, res) {
-        if (err) {
-          if (err.errorCode == 'non-fast-forward') {
-            var forcePushDialog = components.create('yesnodialog', { title: 'Force push?', details: 'The remote branch can\'t be fast-forwarded.' });
-            forcePushDialog.closed.add(function() {
-              if (!forcePushDialog.result()) return callback();
+    promise = this.server.postPromise('/push', pushReq)
+      .catch(function(err) {
+        if (err.errorCode == 'non-fast-forward') {
+          return components.create('yesnodialog', { title: 'Force push?', details: 'The remote branch can\'t be fast-forwarded.' })
+            .publish()
+            .closeThen(function(diag) {
+              if (!diag.result()) return false;
               pushReq.force = true;
-              self.server.post('/push', pushReq, callbackWithRefSet);
-            });
-            programEvents.dispatch({ event: 'request-show-dialog', dialog: forcePushDialog });
-            return true;
-          }
+              return self.server.postPromise('/push', pushReq);
+            })
         }
-        callbackWithRefSet(err, res);
-      });
+      })
   }
+
+  promise
+    .catch(function(err) { callback(err) })
+    .then(function(res) {
+      if (!res) return;
+      var targetNode = self.graph.getNode(target);
+      if (self.graph.checkedOutBranch() == self.refName) {
+        self.graph.HEADref().node(targetNode);
+      }
+      self.node(targetNode);
+      callback();
+    });
 }
 
 RefViewModel.prototype.remove = function(callback) {
   var self = this;
   var url = this.isTag ? '/tags' : '/branches';
   if (this.isRemote) url = '/remote' + url;
-  this.server.del(url, { path: this.graph.repoPath(), remote: this.isRemote ? this.remote : null, name: this.refName }, function(err) {
-    if (!err) {
-      self.node().removeRef(self);
-      self.graph.refsByRefName[self.name] = undefined;
-    }
 
+  return this.server.delPromise(url, { path: this.graph.repoPath(), remote: this.isRemote ? this.remote : null, name: this.refName }).then(function(err) {
```

Comment:
The `err` parameter is not used (at the end of the line). Maybe you want an empty `catch`?